src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -95,14 +99,63 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
 
     return len;
 }
+
+#if !HAVE_IFADDRS
+/* This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+ * might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+ * so that we can dynamically load the getifaddrs function and use it.
+ */
+struct ifaddrs {
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union {
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS
+// try to load the getifaddrs and freeifaddrs functions manually (this is necessary on Android prior to API 24)
+static int (*getifaddrs)(struct ifaddrs**) = NULL;
+static void (*freeifaddrs)(struct ifaddrs*) = NULL;
+
+static bool ensure_getifaddrs_is_loaded() {
+    static bool loading_already_attempted = false;
+
+    if (!loading_already_attempted)
+    {
+        loading_already_attempted = true;
+
+        void *libc = dlopen(LIBC_FILENAME, RTLD_NOW);
+        if (libc)
+        {
+            getifaddrs = (int (*)(struct ifaddrs**)) dlsym(libc, "getifaddrs");
+            freeifaddrs = (void (*)(struct ifaddrs*)) dlsym(libc, "freeifaddrs");
+        }
+    }
+
+    return getifaddrs != NULL && freeifaddrs != NULL;
+}
 #endif
 
 int32_t SystemNative_EnumerateInterfaceAddresses(void* context,
                                                IPv4AddressFound onIpv4Found,
                                                IPv6AddressFound onIpv6Found,
                                                LinkLayerAddressFound onLinkLayerFound)
 {
-#if HAVE_GETIFADDRS
+#if !HAVE_GETIFADDRS
+    if (!ensure_getifaddrs_is_loaded())
```

Comment:
Do we want to do it everywhere or on adroind only?

Reply:
We could make this explicitly Android-only. I thought we could keep this generic by checking if the function is available or not but maybe this could behave unpredictably on some platform with a weird libc that'll be added in the future.

Reply:
@janvorli @jkotas thoughts?

Reply:
I would make it Android only.

I assume this workaround can be deleted once the minimum supported Android API version is API 24 or greater. It may be useful to explicitly say that in a comment. (Just curious - what is the minimum supported Android API version for .NET 7?)

Reply:
MAUI should [support API 21 (Android 5.0)](https://docs.microsoft.com/en-us/dotnet/maui/supported-platforms). I'm not sure if that will change for .NET 7.

Reply:
21 is the min for .NET6.  We haven't discussed it for .NET7 yet.